### DISCUSSION TITLE
How to tell the compiler that something is a ref and can be mutated?

### DISCUSSION THREAD

**Discussion by gaearon** [reactwg/react-compiler#36] [General Questions] [answered]
This is getting flagged by the rule:

```js
export function GalleryItem({
  // ...
  thumbDimsRef,
}: Props) {
  // ...
  return (
    // ...
    <Image
      onLoad={e => {
        // eslint: Mutating component props or hook arguments is not allowed. Consider using a local variable instead
        thumbDimsRef.current[index] = {
          width: e.source.width,
          height: e.source.height,
        }
      }}
    />
```

How do I explain to the compiler that this is safe?

**Comment by gaearon** [comment]
Another question is how to model the "mutable stuff in state" pattern.

```js
  React.useEffect(() => {
    if (state.needsPersist) {
      state.needsPersist = false
      // ...
    }
  }, [state])
```

I know this isn't exactly blessed, but I recall discussions about maybe changing the inner type to a ref to communicate to the compiler that this information isn't used for rendering. However, holding a ref in state and modifying its `.current` triggers the same error for me.

**Reply by poteto** [reply]
If I'm not wrong today the compiler will bailout if the state variable is mutated directly. We do need to discuss this pattern more and whether we support it in the compiler

cc @josephsavona

**Comment by aeharding** [comment, marked as answer]
enableTreatRefLikeIdentifiersAsRefs should work:

https://playground.react.dev/#N4Igzg9grgTgxgUxALhAejQAgAIIHYCGARgDYIAqMCBALgEoIBmAMgJYDWCAkgCb42tGrBDDABBMA0ZgAOnjkIAHgAcIMGpkZQ8cARDyYA4gRJkYATy40EAWwAUwOZkwZMAOg9PMNABZQbRAAirDaSTAA0cgC+yJgACjAQymAAlJiOBi5YHm5eVDSwBnZezq45JZgAPFw2BADmCBXO+swQBDwAvMAImB0AfOlNzlmYCGAkrHg0sQCyUDS0k3WYcBA2qnj8mMqJyZhqmD4QEOyYBDB1-vxgmKw3eBAaJiQQAO4IPG6YAML6YKx8GCYKD-PDLAiYF5wEyYABu51YxDItzwYGs7SGzl8-iCITCjDccFgVCmAG1JnxFABdXqDTLDBmYV4A3yxBBuSCwRBuZk8XyRemM5w+BCsOo+aajDnQeDskViiUCoXOKJDKKqwVoPpeFLREBRIA

**Reply by gaearon** [reply]
Do you know if there's some way to enable this globally? Or are flags per-file?

**Reply by aeharding** [reply]
Here's how I do it globally in my project. New file `compilerOptions.js`:

https://github.com/aeharding/voyager/blob/main/compilerOptions.js

And then I reference in `vite.config.js`:

https://github.com/aeharding/voyager/blob/5993a0d4a03a63783797d9b8a6dce55c74641e34/vite.config.ts#L27

and `eslint.config.js`:

https://github.com/aeharding/voyager/blob/5993a0d4a03a63783797d9b8a6dce55c74641e34/eslint.config.js#L34

**Reply by gaearon** [reply]
Perfect, thanks. @poteto would be nice to document this

**Reply by slorber** [reply]
FYI

![CleanShot 2024-11-18 at 14 06 15](https://github.com/user-attachments/assets/75405910-1f32-4df7-908d-b5febed071dd)

**Reply by poteto** [reply]
Yeah I think we'll want to turn on `enableTreatRefLikeIdentifiersAsRefs` by default, we need to test it more internally first however. Our general hope with compiler flags is that these are not really meant to be configured, they're more like experimentation flags than configuration
